test/service_test.cc

```diff
@@ -0,0 +1,458 @@
+      struct ServiceDataStructure::Chirp chirp;
+      // ServiceDataStructure::ReturnCodes
+      auto ret = service_data_structure_.ReadChirp(id, &chirp);
+      // Reading should be successful
+      ASSERT_EQ(ServiceDataStructure::OK, ret);
+      EXPECT_EQ(last_id, chirp.parent_id);
+      last_id = chirp.id;
+
+      // colect the texts posted
+      chirps_content_from_backend.push_back(chirp.text);
+    }
+    // Check if the contents are identical
+    EXPECT_EQ(chirps_content, chirps_content_from_backend); 
+
+    // Test Editing
+    chirps_content_from_backend.clear();
+    for(size_t j = 0; j < chirp_ids.size(); ++j) {
+      // ServiceDataStructure::ReturnCodes
+      auto ret = session->EditChirp(chirp_ids[j], chirps_content_after_edit[j]);
+      // Editing should be successful
+      ASSERT_EQ(ServiceDataStructure::OK, ret);
+    }
+
+    // Read from backend
+    // to see if the results are identical
+    for(auto& id : session->SessionGetUserChirpList()) {
+      struct ServiceDataStructure::Chirp chirp;
+      // ServiceDataStructure::ReturnCodes
+      auto ret = service_data_structure_.ReadChirp(id, &chirp);
+      // Reading should be successful
+      ASSERT_EQ(ServiceDataStructure::OK, ret);
+
+      // colect the texts posted
+      chirps_content_from_backend.push_back(chirp.text);
+    }
+    // Check if the contents are identical
+    EXPECT_EQ(chirps_content_after_edit, chirps_content_from_backend);
+
+    // Test Deleting
+    chirps_content_from_backend.clear();
+    for(size_t j = 0; j < kHalfTestCase; ++j) {
+      // ServiceDataStructure::ReturnCodes
+      auto ret = session->DeleteChirp(chirp_ids[j]);
+      // Deleting should be successful
+      EXPECT_EQ(ServiceDataStructure::OK, ret);
+      ret = session->DeleteChirp(chirp_ids[j]);
+      // Deleting should be unsuccessful since this chirp is already deleted
+      EXPECT_EQ(ServiceDataStructure::CHIRP_ID_NOT_FOUND, ret);
+    }
+
+    // Read from backend
+    // to see if the results are identical
+    for(auto& id : session->SessionGetUserChirpList()) {
+      struct ServiceDataStructure::Chirp chirp;
+      // ServiceDataStructure::ReturnCodes
+      auto ret = service_data_structure_.ReadChirp(id, &chirp);
+      // Reading should be successful
+      ASSERT_EQ(ServiceDataStructure::OK, ret);
+
+      // colect the texts posted
+      chirps_content_from_backend.push_back(chirp.text);
+    }
+    // Check if the contents are identical
+    EXPECT_EQ(chirps_content_after_delete, chirps_content_from_backend);
+  }
+}
+
+// This tests on `Follow` and `MonitorFrom` in `ServiceDataStructure`
+TEST_F(ServiceTestDataStructure, FollowAndMonitorTest) {
+  // Make each user follows the next user
+  for(size_t i = 0; i < kNumOfUsersPreset; ++i) {
+    // std::unique_ptr<ServiceDataStructure::UserSession>
+    auto session = service_data_structure_.UserLogin(user_list_[i]);
+    // Login should be successful
+    EXPECT_NE(nullptr, session);
```

Comment:
I think this should be ASSERT_NE because if user can't login then the follow will fail

Reply:
Follows for EXPECTs below